sql: implement ConstructFilter and ConstructExplain in the new factory

yuzefovich opened the pull request:
**sql: implement ConstructFilter in the new factory**

This commit implements `ConstructFilter` in the new factory by simply
reusing `AddFilter` function which attempts to push the filter into the
post-processing stage of the last stage of processors.

This commit does make an important decision: it introduces the notion of
"partially distributed" plans (meaning that some stages of the processors
can be distributed whereas other stages are not). As an example, if the
query contains a distsql-blacklisted function, the old code will not
distribute the whole plan, but with this approach we might distribute
scans, then bring the streams together to the gateway, and plan everything
after and including the blacklisted function locally. We will be taking
this even further eventually - if it is beneficial, we can plan a later
stage in distributed fashion (but the new factory doesn't yet implement
methods that would result in such behavior).

Release note: None

**sql: introduce ternary planDistribution struct**

In the previous commit we introduced a notion of "partially distributed"
plans, and this commit introduces a ternary `planDistribution` struct to
represent all possible plan distributions: "local", "partial", and
"full". `queryMeta` adjustment is left as TODO.

Release note: None

**sql: update 'distributed' field of EXPLAIN output**

In the previous commits we introduced a notion of "partially distributed"
plans, and this commit adjusts the EXPLAIN code accordingly.

Release note (sql change): The output of some EXPLAIN variants has been
changed so that "distributed" field is now named "distribution" and can
be of "local", "partial", and "full" states ("partial" is currently only
possible with experimental setting).

**sql: implement ConstructExplain in distSQLSpecExecFactory**

This commit adds an implementation of `ConstructExplain` to
`distSQLSpecExecFactory` that is shared with `execFactory`. This
introduces a notion of "partially physical" plan (meaning that we have
a `planNode` at the root and an already completely planned physically
leaf. The intention is to be able to test physical planning of the new
factory. As a result of this approach, `EXPLAIN (DISTSQL)` and `EXPLAIN
(VEC)` work well but "vanilla" `EXPLAIN` (plan) doesn't. The reason for
that is the vanilly explain walks over the plan and populates the
"attributes"; however, now we have a mixed planNode-spec representation
which existing walker doesn't support. I see two ways to address this:
either fully fall back to the old factory for the vanilla explain or
implement a separate walker on DistSQL specs. I'm leaning towards the
latter option which is left as TODO.

Release note: None

yuzefovich committed:
sql: implement ConstructFilter in the new factory

This commit implements `ConstructFilter` in the new factory by simply
reusing `AddFilter` function which attempts to push the filter into the
post-processing stage of the last stage of processors.

This commit does make an important decision: it introduces the notion of
"partially distributed" plans (meaning that some stages of the processors
can be distributed whereas other stages are not). As an example, if the
query contains a distsql-blacklisted function, the old code will not
distribute the whole plan, but with this approach we might distribute
scans, then bring the streams together to the gateway, and plan everything
after and including the blacklisted function locally. We will be taking
this even further eventually - if it is beneficial, we can plan a later
stage in distributed fashion (but the new factory doesn't yet implement
methods that would result in such behavior).

Release note: None

yuzefovich committed:
sql: introduce ternary planDistribution struct

In the previous commit we introduced a notion of "partially distributed"
plans, and this commit introduces a ternary `planDistribution` struct to
represent all possible plan distributions: "local", "partial", and
"full". `queryMeta` adjustment is left as TODO.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/49857)
<!-- Reviewable:end -->

yuzefovich committed:
sql: update 'distributed' field of EXPLAIN output

In the previous commits we introduced a notion of "partially distributed"
plans, and this commit adjusts the EXPLAIN code accordingly.

Release note (sql change): The output of some EXPLAIN variants has been
changed so that "distributed" field is now named "distribution" and can
be of "local", "partial", and "full" states ("partial" is currently only
possible with experimental setting).

yuzefovich reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/49857)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @asubiotto and @rytaft)

___
*[pkg/sql/opt_exec_factory.go, line 234 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/49857#-M9-vBdS9bX9_ELBz2zB:-M9-vBdS9bX9_ELBz2zC:b-ns3nd3) ([raw file](https://github.com/cockroachdb/cockroach/blob/c0c55c14888fd14947bb3b2b9c217cd4f3e62b6d/pkg/sql/opt_exec_factory.go#L234)):*
> ```Go
> 		s.filter = s.filterVars.Rebind(filter, true /* alsoReset */, false /* normalizeToNonNil */)
> 		// Note: if the filter statically evaluates to true, s.filter stays nil.
> 		// TODO(yuzefovich): I think this is unnecessary - ConstructScan should
> ```

cc @rytaft 


<!-- Sent from Reviewable.io -->

yuzefovich commented:
First four commits are in a separate PR, but I wanted to get early feedback about the introduction of "partially distributed" plans in the fifth commit. cc @jordanlewis @RaduBerinde

yuzefovich committed:
sql: implement ConstructExplain in distSQLSpecExecFactory

This commit adds an implementation of `ConstructExplain` to
`distSQLSpecExecFactory` that is shared with `execFactory`. This
introduces a notion of "partially physical" plan (meaning that we have
a `planNode` at the root and an already completely planned physically
leaf. The intention is to be able to test physical planning of the new
factory. As a result of this approach, `EXPLAIN (DISTSQL)` and `EXPLAIN
(VEC)` work well but "vanilla" `EXPLAIN` (plan) doesn't. The reason for
that is the vanilla EXPLAIN walks over the plan and populates the
"attributes"; however, now we have a mixed planNode-DistSQL-spec
representation which existing walker doesn't support. I see two ways to
address this: either fully fall back to the old factory for the vanilla
EXPLAIN or implement a separate walker on DistSQL specs. I'm leaning
towards the latter option which is left as TODO.

Release note: None

RaduBerinde commented:
I am thinking of ways to revamp Explain. Instead of building a planNode tree and walking it, I think EXPLAIN (PLAN) should more or less be a "trace" of the exec factory calls. One way I thought about was to have an exec.Factory method that returns a different "explain factory" implementation which is used inside the explain; it simply records whatever we want to record from the calls. I'd spend as little energy on EXPLAIN (PLAN) right now, as hopefully it will become easier in that new world.

yuzefovich commented:
RFAL. The fourth commit adds a couple of tests of the new planning, and I intend to add more in a separate PR.

> I'd spend as little energy on EXPLAIN (PLAN) right now

Ok, this makes sense, thanks!

asubiotto reviewed (changes requested):
> Reviewed 3 of 4 files at r5, 2 of 2 files at r6, 10 of 10 files at r7, 60 of 60 files at r8, 8 of 8 files at r9.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/49857)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @rytaft and @yuzefovich)

___
*[pkg/sql/conn_executor_exec.go, line 760 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/49857#-M9OHI878UwaIKDdzTab:-M9OHI878UwaIKDdzTac:b-upx7zh) ([raw file](https://github.com/cockroachdb/cockroach/blob/ec4a129719c65296923d13cdc0b1f4705c4a41b8/pkg/sql/conn_executor_exec.go#L760)):*
> ```Go
> 
> 	ex.sessionTracing.TracePlanCheckStart(ctx)
> 	distribution := willDistributePlan(
> ```

We probably should rename this now that it doesn't return a boolean (maybe to `getPlanDistribution` or something similar)
___
*[pkg/sql/conn_executor_exec.go, line 779 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/49857#-M9OHT5N45orX1JZQnBH:-M9OHT5N45orX1JZQnBI:b-ruq67q) ([raw file](https://github.com/cockroachdb/cockroach/blob/ec4a129719c65296923d13cdc0b1f4705c4a41b8/pkg/sql/conn_executor_exec.go#L779)):*
> ```Go
> 	}
> 	queryMeta.phase = executing
> 	// TODO(yuzefovich): introduce ternary planDistribution into queryMeta.
> ```

Why not do this in this commit?
___
*[pkg/sql/distsql_physical_planner.go, line 259 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/49857#-M9OHYCz6bIUDylaxb7V:-M9OHYD-BHRspGhrfYpg:b-w8jfiy) ([raw file](https://github.com/cockroachdb/cockroach/blob/ec4a129719c65296923d13cdc0b1f4705c4a41b8/pkg/sql/distsql_physical_planner.go#L259)):*
> ```Go
> 
> 	// partiallyDistributedPlan indicates that some parts of the plan are
> 	// distributed while other parts are not (due to limitations of DistSQL).
> ```

Add example
___
*[pkg/sql/distsql_physical_planner.go, line 281 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/49857#-M9OHccJER9Cij32AcYl:-M9OHccJER9Cij32AcYm:b-326zmr) ([raw file](https://github.com/cockroachdb/cockroach/blob/ec4a129719c65296923d13cdc0b1f4705c4a41b8/pkg/sql/distsql_physical_planner.go#L281)):*
> ```Go
> 		return "full"
> 	default:
> 		panic("unsupported planDistribution")
> ```

fmt.Sprintf the unsupported value
___
*[pkg/sql/distsql_running.go, line 870 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/49857#-M9OHmxr1u_FB8-E0kK_:-M9OHmxr1u_FB8-E0kKa:bufeugk) ([raw file](https://github.com/cockroachdb/cockroach/blob/ec4a129719c65296923d13cdc0b1f4705c4a41b8/pkg/sql/distsql_running.go#L870)):*
> ```Go
> 	var distributeSubquery bool
> 	if maybeDistribute {
> 		subqueryDistribution := willDistributePlan(
> ```

Can you collapse this call and elsewhere to `willDistributePlan(<...>).willDistribute()`
___
*[pkg/sql/distsql_spec_exec_factory.go, line 32 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/49857#-M9ODGSnEhyWqnjYbl-k:-M9ODGSnEhyWqnjYbl-l:b-2mbs54) ([raw file](https://github.com/cockroachdb/cockroach/blob/7aab3b0c352b1b23f073b54561aca7d988d4bccf/pkg/sql/distsql_spec_exec_factory.go#L32)):*
> ```Go
> 	dsp     *DistSQLPlanner
> 	// planContexts is a utility struct that stores already instantiated
> 	// planning contexts. It should not be accessed directly, use getPlanCtx()
> ```

I think this comment needs to also explain the why of this struct, not just the what 
___
*[pkg/sql/distsql_spec_exec_factory.go, line 45 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/49857#-M9OCwXF9HH2lBSJzN3R:-M9OCwXG-mG9xzIh7arZ:bdajv6) ([raw file](https://github.com/cockroachdb/cockroach/blob/7aab3b0c352b1b23f073b54561aca7d988d4bccf/pkg/sql/distsql_spec_exec_factory.go#L45)):*
> ```Go
> 
> func newDistSQLSpecExecFactory(p *planner) exec.Factory {
> 	_, singleTenant := p.execCfg.NodeID.OptionalNodeID()
> ```

Use `p.execCfg.Codec.ForSystemTenant()` instead
___
*[pkg/sql/distsql_spec_exec_factory.go, line 222 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/49857#-M9OEk_k0qyym0RVddfg:-M9OEk_k0qyym0RVddfh:bmv9sp2) ([raw file](https://github.com/cockroachdb/cockroach/blob/7aab3b0c352b1b23f073b54561aca7d988d4bccf/pkg/sql/distsql_spec_exec_factory.go#L222)):*
> ```Go
> 			// single node. We could do so on one of the nodes that streams
> 			// originate from, but for now we choose the gateway.
> 			gatewayNodeID := e.dsp.nodeDesc.NodeID
> ```

Hmm, I don't think we should introduce new uses of `NodeID`. You probably want to use `p.execCfg.NodeID.SQLInstanceID` instead. cc @tbg I think we probably need to remove this direct reference to `nodeDesc` in the distsql physical planner since that won't be available in a tenant, right?
___
*[pkg/sql/distsql_spec_exec_factory.go, line 236 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/49857#-M9OFf8n1VEwoBtDBlNs:-M9OFf8o3B1Z3DYnfXiy:b-ns3nd3) ([raw file](https://github.com/cockroachdb/cockroach/blob/7aab3b0c352b1b23f073b54561aca7d988d4bccf/pkg/sql/distsql_spec_exec_factory.go#L236)):*
> ```Go
> 	// last stage if there are RenderExprs present. I think that optbuilder
> 	// should be the one pushing the filters through renders, so we don't
> 	// actually need to do anything about it, right?
> ```

cc @rytaft 
___
*[pkg/sql/distsql_spec_exec_factory.go, line 212 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/49857#-M9OIA6MELpY6rI3WDpV:-M9OIA6MELpY6rI3WDpW:bwlilqy) ([raw file](https://github.com/cockroachdb/cockroach/blob/ec4a129719c65296923d13cdc0b1f4705c4a41b8/pkg/sql/distsql_spec_exec_factory.go#L212)):*
> ```Go
> 	)
> 
> 	return planMaybePhysical{physPlan: &p, distribution: distribution}, err
> ```

It doesn't feel right to have a `distribution` field here coupled with the `planCtx`. Can we update the `planCtx` to have this new distribution field and keep that information there?
___
*[pkg/sql/distsql_spec_exec_factory.go, line 220 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/49857#-M9OIZR_3qXiR8l1Y5UR:-M9OIZR_3qXiR8l1Y5US:bw2lbb0) ([raw file](https://github.com/cockroachdb/cockroach/blob/ec4a129719c65296923d13cdc0b1f4705c4a41b8/pkg/sql/distsql_spec_exec_factory.go#L220)):*
> ```Go
> 	plan := n.(planMaybePhysical)
> 	physPlan := plan.physPlan
> 	distributedLastStage := len(physPlan.ResultRouters) > 1
> ```

Should we extract this logic into a helper method on `PhysicalPlan`? I think it might be nicer to read this way.
___
*[pkg/sql/exec_util.go, line 918 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/49857#-M9OLMMx1BTmkxPbLw4f:-M9OLMMx1BTmkxPbLw4g:bcx5jab) ([raw file](https://github.com/cockroachdb/cockroach/blob/251099f29932c47fdc3a2cd071a7554534779910/pkg/sql/exec_util.go#L918)):*
> ```Go
> 			if valuesSpec := plan.physPlan.Processors[0].Spec.Core.Values; valuesSpec != nil {
> 				if valuesSpec.NumRows == 0 {
> 					return localPlan
> ```

Could you walk me through how we still return `localPlan` in these cases with the new code?
___
*[pkg/sql/exec_util.go, line 899 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/49857#-M9OLF_s4-XVWU3vhRGN:-M9OLF_s4-XVWU3vhRGO:b-mlvepx) ([raw file](https://github.com/cockroachdb/cockroach/blob/c4fb8212f80434e77fae245c50cbf2411bbe3c94/pkg/sql/exec_util.go#L899)):*
> ```Go
> // planning being handled by DistSQL physical planner. If the passed-in plan
> // already has physical representation (meaning it was constructed by a
> // distSQLSpecExecFactory), then the pre-populated plan distribution is
> ```

nit: should this be in the previous commit?
___
*[pkg/sql/plan_opt.go, line 207 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/49857#-M9OLcBKDrMqRgTTHCH6:-M9OLcBKDrMqRgTTHCH7:bqn0w3y) ([raw file](https://github.com/cockroachdb/cockroach/blob/c4fb8212f80434e77fae245c50cbf2411bbe3c94/pkg/sql/plan_opt.go#L207)):*
> ```Go
> 			// TODO(yuzefovich): make the logging conditional on the verbosity
> 			// level once new DistSQL planning is no longer experimental.
> 			log.Infof(ctx, "%v", err)
> ```

Do you want to add more context to the log message or is this fine?


<!-- Sent from Reviewable.io -->

tbg reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/49857)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @asubiotto, @rytaft, and @yuzefovich)

___
*[pkg/sql/distsql_spec_exec_factory.go, line 222 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/49857#-M9OEk_k0qyym0RVddfg:-M9OMQnM62BMquL6Ft95:b-wf8ifw) ([raw file](https://github.com/cockroachdb/cockroach/blob/7aab3b0c352b1b23f073b54561aca7d988d4bccf/pkg/sql/distsql_spec_exec_factory.go#L222)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

Hmm, I don't think we should introduce new uses of `NodeID`. You probably want to use `p.execCfg.NodeID.SQLInstanceID` instead. cc @tbg I think we probably need to remove this direct reference to `nodeDesc` in the distsql physical planner since that won't be available in a tenant, right?
</blockquote></details>

Correct, right now I'm hard-coding some bogus values there. We should not be picking up new references or if we do, make sure they're not breaking multi-tenancy (i.e. they don't matter for local flows).


<!-- Sent from Reviewable.io -->

rytaft reviewed (approved):
> Reviewed 3 of 4 files at r5, 2 of 2 files at r6, 10 of 10 files at r7, 10 of 60 files at r8, 8 of 8 files at r9.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/49857)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @asubiotto and @yuzefovich)

___
*[pkg/sql/distsql_spec_exec_factory.go, line 236 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/49857#-M9OFf8n1VEwoBtDBlNs:-M9O_iHP5Kr3BHHg2Xf7:byx4gef) ([raw file](https://github.com/cockroachdb/cockroach/blob/7aab3b0c352b1b23f073b54561aca7d988d4bccf/pkg/sql/distsql_spec_exec_factory.go#L236)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

cc @rytaft 
</blockquote></details>

Yea, the optimizer will always push `Select` below `Project` if possible (I think that means the same thing as pushing filters through renders, right?).
___
*[pkg/sql/opt_exec_factory.go, line 234 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/49857#-M9-vBdS9bX9_ELBz2zB:-M9Oab2CBkbqLFgXLZbl:b-lmvgkn) ([raw file](https://github.com/cockroachdb/cockroach/blob/c0c55c14888fd14947bb3b2b9c217cd4f3e62b6d/pkg/sql/opt_exec_factory.go#L234)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

cc @rytaft 
</blockquote></details>

The required ordering should be the same if the filter is directly above the scan.


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/49857)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @asubiotto and @rytaft)

___
*[pkg/sql/conn_executor_exec.go, line 760 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/49857#-M9OHI878UwaIKDdzTab:-M9RNrD02VgVKbk1mpa_:bfbq99i) ([raw file](https://github.com/cockroachdb/cockroach/blob/ec4a129719c65296923d13cdc0b1f4705c4a41b8/pkg/sql/conn_executor_exec.go#L760)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

We probably should rename this now that it doesn't return a boolean (maybe to `getPlanDistribution` or something similar)
</blockquote></details>

Good point, done.
___
*[pkg/sql/conn_executor_exec.go, line 779 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/49857#-M9OHT5N45orX1JZQnBH:-M9RP8XkDxklyfIAHsKt:bmm2162) ([raw file](https://github.com/cockroachdb/cockroach/blob/ec4a129719c65296923d13cdc0b1f4705c4a41b8/pkg/sql/conn_executor_exec.go#L779)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

Why not do this in this commit?
</blockquote></details>

Because `queryMeta.isDistributed` eventually is used to assign to a field in a protobuf that is used to populate some internal virtual tables. I didn't try changing that, but I'm worried the diff of the test output will be big and deserves to be in a separate commit. Also, for the time being it seems acceptable to treat "partially distributed" plans as "distributed" for the purposes of the query metadata.
___
*[pkg/sql/distsql_physical_planner.go, line 259 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/49857#-M9OHYCz6bIUDylaxb7V:-M9RQniC5daLfPrE9NkX:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ec4a129719c65296923d13cdc0b1f4705c4a41b8/pkg/sql/distsql_physical_planner.go#L259)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

Add example
</blockquote></details>

Done.
___
*[pkg/sql/distsql_physical_planner.go, line 281 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/49857#-M9OHccJER9Cij32AcYl:-M9RQvCdE6YRdRX_Ihv8:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ec4a129719c65296923d13cdc0b1f4705c4a41b8/pkg/sql/distsql_physical_planner.go#L281)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

fmt.Sprintf the unsupported value
</blockquote></details>

Done.
___
*[pkg/sql/distsql_running.go, line 870 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/49857#-M9OHmxr1u_FB8-E0kK_:-M9RRHX2-HSOEufq2QyD:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ec4a129719c65296923d13cdc0b1f4705c4a41b8/pkg/sql/distsql_running.go#L870)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

Can you collapse this call and elsewhere to `willDistributePlan(<...>).willDistribute()`
</blockquote></details>

Done.
___
*[pkg/sql/distsql_spec_exec_factory.go, line 32 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/49857#-M9ODGSnEhyWqnjYbl-k:-M9RJYgfEbzFId37nKao:b35wl8g) ([raw file](https://github.com/cockroachdb/cockroach/blob/7aab3b0c352b1b23f073b54561aca7d988d4bccf/pkg/sql/distsql_spec_exec_factory.go#L32)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

I think this comment needs to also explain the why of this struct, not just the what 
</blockquote></details>

Expanded the comment.
___
*[pkg/sql/distsql_spec_exec_factory.go, line 45 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/49857#-M9OCwXF9HH2lBSJzN3R:-M9RJmDjBlDTdki8KcZx:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/7aab3b0c352b1b23f073b54561aca7d988d4bccf/pkg/sql/distsql_spec_exec_factory.go#L45)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

Use `p.execCfg.Codec.ForSystemTenant()` instead
</blockquote></details>

Done.
___
*[pkg/sql/distsql_spec_exec_factory.go, line 222 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/49857#-M9OEk_k0qyym0RVddfg:-M9RKnbv9w9kAm1QFs38:bfbryk1) ([raw file](https://github.com/cockroachdb/cockroach/blob/7aab3b0c352b1b23f073b54561aca7d988d4bccf/pkg/sql/distsql_spec_exec_factory.go#L222)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

Correct, right now I'm hard-coding some bogus values there. We should not be picking up new references or if we do, make sure they're not breaking multi-tenancy (i.e. they don't matter for local flows).
</blockquote></details>

Well, `AddSingleGroupStage` method already exists, and it asks for `NodeID`, what should I give to it then?

This usage of `NodeID` doesn't break the multitenancy because I simply need a node on which to plan a single processor, and the node can be arbitrary. So far we've had the concept of a "gateway" node, but it seems like multitenancy initiative will remove that concept. I think during that removal is right time to change the usage as well.
___
*[pkg/sql/distsql_spec_exec_factory.go, line 236 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/49857#-M9OFf8n1VEwoBtDBlNs:-M9RLfTp4tSZYm-yazMj:bkuvnoj) ([raw file](https://github.com/cockroachdb/cockroach/blob/7aab3b0c352b1b23f073b54561aca7d988d4bccf/pkg/sql/distsql_spec_exec_factory.go#L236)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

Yea, the optimizer will always push `Select` below `Project` if possible (I think that means the same thing as pushing filters through renders, right?).
</blockquote></details>

The situation is whether the optimizer is smart enough to change `ConstructFilter -> ConstructProject -> ConstructScan` into `ConstructProject -> ConstructFilter -> ConstructScan`. I understand that it is not always possible (when the filter actually uses the result of projection), and there is nothing we could do, but in other cases the optimizer should be able to push it. I removed the TODO.
___
*[pkg/sql/distsql_spec_exec_factory.go, line 212 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/49857#-M9OIA6MELpY6rI3WDpV:-M9RRQue8ZsHRzH6HpHs:b-edfapq) ([raw file](https://github.com/cockroachdb/cockroach/blob/ec4a129719c65296923d13cdc0b1f4705c4a41b8/pkg/sql/distsql_spec_exec_factory.go#L212)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

It doesn't feel right to have a `distribution` field here coupled with the `planCtx`. Can we update the `planCtx` to have this new distribution field and keep that information there?
</blockquote></details>

What do you mean "coupled with the `planCtx`"? What happens is that `planMaybePhysical.distribution` describes the distribution of the plan that we have built so far, and we're going to be updating the field with every `Construct*` call. `planCtx` is a temporary object used to provide some information that is needed during the physical planning of a single stage of processors. So two things have different scope: `distribution` is "global" whereas `planCtx` is "local". Does it make sense?
___
*[pkg/sql/distsql_spec_exec_factory.go, line 220 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/49857#-M9OIZR_3qXiR8l1Y5UR:-M9RSvrE4nkNVMruGRlM:bfbq99i) ([raw file](https://github.com/cockroachdb/cockroach/blob/ec4a129719c65296923d13cdc0b1f4705c4a41b8/pkg/sql/distsql_spec_exec_factory.go#L220)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

Should we extract this logic into a helper method on `PhysicalPlan`? I think it might be nicer to read this way.
</blockquote></details>

Good point, done.
___
*[pkg/sql/exec_util.go, line 918 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/49857#-M9OLMMx1BTmkxPbLw4f:-M9RV5Q61O7viKqzr66k:b-f3x9ic) ([raw file](https://github.com/cockroachdb/cockroach/blob/251099f29932c47fdc3a2cd071a7554534779910/pkg/sql/exec_util.go#L918)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

Could you walk me through how we still return `localPlan` in these cases with the new code?
</blockquote></details>

As I described in another comment, with the new factory we're building the DistSQL spec "tree" and `planDistribution` iteratively at the same time while traversing the `opt.Expr` tree. So by the time we get to this method if we have physical representation of the plan, then `distribution` field must have been populated accordingly.

At the moment, these "empty" nodes are not constructed by the new factory, but the rough outline of how the plan distribution is determined in there is: for every `Construct*` method we have a recommendation on whether we want to distribute that particular `opt.Expr` given the distribution of the last stage of the physical plan so far, then based on that `recommendation` we get either local or distributed `planCtx` which will determine whether `DistSQLPhysicalPlanner` methods will plan locally or in a distributed fashion; now we might have added a new stage of the processors, so we update `planMaybePhysical.distribution` field accordingly.
___
*[pkg/sql/exec_util.go, line 899 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/49857#-M9OLF_s4-XVWU3vhRGN:-M9RUtNBDjVYgQ5AA_X0:b-xew2je) ([raw file](https://github.com/cockroachdb/cockroach/blob/c4fb8212f80434e77fae245c50cbf2411bbe3c94/pkg/sql/exec_util.go#L899)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

nit: should this be in the previous commit?
</blockquote></details>

Yep, updated.
___
*[pkg/sql/opt_exec_factory.go, line 234 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/49857#-M9-vBdS9bX9_ELBz2zB:-M9RND3R63SLpDuFzm2L:bg5s9qb) ([raw file](https://github.com/cockroachdb/cockroach/blob/c0c55c14888fd14947bb3b2b9c217cd4f3e62b6d/pkg/sql/opt_exec_factory.go#L234)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

The required ordering should be the same if the filter is directly above the scan.
</blockquote></details>

Cool, removed the line and left a comment.
___
*[pkg/sql/plan_opt.go, line 207 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/49857#-M9OLcBKDrMqRgTTHCH6:-M9RUs9rBeJmYh0hW23q:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/c4fb8212f80434e77fae245c50cbf2411bbe3c94/pkg/sql/plan_opt.go#L207)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

Do you want to add more context to the log message or is this fine?
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

asubiotto reviewed (commented):
> Reviewed 74 of 74 files at r10, 11 of 11 files at r11, 60 of 60 files at r12.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/49857)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @asubiotto, @rytaft, @tbg, and @yuzefovich)

___
*[pkg/sql/distsql_spec_exec_factory.go, line 222 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/49857#-M9OEk_k0qyym0RVddfg:-M9SNLH8AHZrg8ENJucq:b-45522w) ([raw file](https://github.com/cockroachdb/cockroach/blob/7aab3b0c352b1b23f073b54561aca7d988d4bccf/pkg/sql/distsql_spec_exec_factory.go#L222)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

Well, `AddSingleGroupStage` method already exists, and it asks for `NodeID`, what should I give to it then?

This usage of `NodeID` doesn't break the multitenancy because I simply need a node on which to plan a single processor, and the node can be arbitrary. So far we've had the concept of a "gateway" node, but it seems like multitenancy initiative will remove that concept. I think during that removal is right time to change the usage as well.
</blockquote></details>

For now I think we should just cast to a node ID until we remove references to node IDs in the planning code.

The thing is that in a multitenant environment, this `NodeID` is not going to be anything useful (it's not available). Multitenancy is not going to remove the concept of a gateway node, just replace the concept of a NodeID that we use to identify different nodes with SQLInstanceIDs. It'll serve the same purpose for identification, just explicitly target SQL nodes (NodeID will be a KV concept) .

By breaking multitenancy, I think @tbg just means that we can't access this code path if running a multitenant instance which is true in this case because of the fact we'll only run single-node flows (for the first "phase" of multitenancy) and we check `physPlan.ResultRouters` above.
___
*[pkg/sql/distsql_spec_exec_factory.go, line 212 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/49857#-M9OIA6MELpY6rI3WDpV:-M9SPdii73KECPdUnfeR:b-ciyy6i) ([raw file](https://github.com/cockroachdb/cockroach/blob/ec4a129719c65296923d13cdc0b1f4705c4a41b8/pkg/sql/distsql_spec_exec_factory.go#L212)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

What do you mean "coupled with the `planCtx`"? What happens is that `planMaybePhysical.distribution` describes the distribution of the plan that we have built so far, and we're going to be updating the field with every `Construct*` call. `planCtx` is a temporary object used to provide some information that is needed during the physical planning of a single stage of processors. So two things have different scope: `distribution` is "global" whereas `planCtx` is "local". Does it make sense?
</blockquote></details>

I see! 


<!-- Sent from Reviewable.io -->

rytaft reviewed (approved):
> Reviewed 7 of 11 files at r11, 4 of 60 files at r12, 8 of 8 files at r13.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/49857)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @tbg and @yuzefovich)

___
*[pkg/sql/opt_exec_factory.go, line 234 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/49857#-M9-vBdS9bX9_ELBz2zB:-M9Swqsr1S-xXZ8IsJmC:blvk0ak) ([raw file](https://github.com/cockroachdb/cockroach/blob/c0c55c14888fd14947bb3b2b9c217cd4f3e62b6d/pkg/sql/opt_exec_factory.go#L234)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

Cool, removed the line and left a comment.
</blockquote></details>

I'm not totally familiar with how this works in DistSQL, though -- even if the required orderings are the same, isn't it possible that you'll need to enforce the ordering if the scans from different nodes are being combined?

I'd be hesitant about removing this line...


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/49857)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @asubiotto and @rytaft)

___
*[pkg/sql/distsql_spec_exec_factory.go, line 222 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/49857#-M9OEk_k0qyym0RVddfg:-M9WrBL16gFENTL1PYfw:b513p10) ([raw file](https://github.com/cockroachdb/cockroach/blob/7aab3b0c352b1b23f073b54561aca7d988d4bccf/pkg/sql/distsql_spec_exec_factory.go#L222)):*
> For now I think we should just cast to a node ID

Cast what? `SQLInstanceID`? I don't see how I would do so. I'm sorry, but I'm not following this particular suggestion (I did get the general description).

Also I'm not sure why we're paying this much attention to this particular line - I'm not introducing any new functions that are using soon-to-be-removed `NodeID` - I'm merely using an already existing one a single time. I think it should be ok to pass in whatever in order to make the function happy, and this code should be refactored when that function is refactored. It doesn't add much overhead for the multitenancy work, and I think you guys have more context on what needs to be passed in. Maybe I'm missing something though.
___
*[pkg/sql/distsql_spec_exec_factory.go, line 212 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/49857#-M9OIA6MELpY6rI3WDpV:-M9WmwEdAtGEGd1UT9sr:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ec4a129719c65296923d13cdc0b1f4705c4a41b8/pkg/sql/distsql_spec_exec_factory.go#L212)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

I see! 
</blockquote></details>

Done.
___
*[pkg/sql/opt_exec_factory.go, line 234 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/49857#-M9-vBdS9bX9_ELBz2zB:-M9Wncch4gpa_rsCN9yn:bwndhzg) ([raw file](https://github.com/cockroachdb/cockroach/blob/c0c55c14888fd14947bb3b2b9c217cd4f3e62b6d/pkg/sql/opt_exec_factory.go#L234)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

I'm not totally familiar with how this works in DistSQL, though -- even if the required orderings are the same, isn't it possible that you'll need to enforce the ordering if the scans from different nodes are being combined?

I'd be hesitant about removing this line...
</blockquote></details>

Ok, added the line back. Note this is the old `execFactory`, and I think this line is not necessary.

However, your comment made me realize that I'm ignoring `reqOrdering` parameter in the new factory, fixed that.


<!-- Sent from Reviewable.io -->

rytaft reviewed (approved):
> Reviewed 7 of 11 files at r15, 10 of 60 files at r16, 8 of 8 files at r17.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/49857)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @asubiotto)

___
*[pkg/sql/opt_exec_factory.go, line 234 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/49857#-M9-vBdS9bX9_ELBz2zB:-M9YKc5O1ji13Z46i0VZ:bp4uaio) ([raw file](https://github.com/cockroachdb/cockroach/blob/c0c55c14888fd14947bb3b2b9c217cd4f3e62b6d/pkg/sql/opt_exec_factory.go#L234)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

Ok, added the line back. Note this is the old `execFactory`, and I think this line is not necessary.

However, your comment made me realize that I'm ignoring `reqOrdering` parameter in the new factory, fixed that.
</blockquote></details>

Sounds good.

cc @raduberinde who would know for sure about this code since he added/updated this recently.


<!-- Sent from Reviewable.io -->

asubiotto reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/49857#-:-M9YTrxo1dEuUS3rhByK:bg7igkv)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @asubiotto, @tbg, and @yuzefovich)

___
*[pkg/sql/distsql_spec_exec_factory.go, line 222 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/49857#-M9OEk_k0qyym0RVddfg:-M9YS_-16RPa15t-ME8c:bwrumuz) ([raw file](https://github.com/cockroachdb/cockroach/blob/7aab3b0c352b1b23f073b54561aca7d988d4bccf/pkg/sql/distsql_spec_exec_factory.go#L222)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

> For now I think we should just cast to a node ID

Cast what? `SQLInstanceID`? I don't see how I would do so. I'm sorry, but I'm not following this particular suggestion (I did get the general description).

Also I'm not sure why we're paying this much attention to this particular line - I'm not introducing any new functions that are using soon-to-be-removed `NodeID` - I'm merely using an already existing one a single time. I think it should be ok to pass in whatever in order to make the function happy, and this code should be refactored when that function is refactored. It doesn't add much overhead for the multitenancy work, and I think you guys have more context on what needs to be passed in. Maybe I'm missing something though.
</blockquote></details>

This is what I mean:
```
gatewayNodeID := roachpb.NodeID(p.execCfg.NodeID.SQLInstanceID())
```
Similar to what I did in `explain_distsql.go`:
```
		// TODO(asubiotto): This cast from SQLInstanceID to NodeID is temporary:
		//  https://github.com/cockroachdb/cockroach/issues/49596
		flows := physPlan.GenerateFlowSpecs(roachpb.NodeID(params.extendedEvalCtx.NodeID.SQLInstanceID()))
```

It's a fair point that the `nodeDesc` already exists and this will just add an extra usage. I'll leave it up to you to decide what you want to do.


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
TFTRs!

bors r+

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/49857#-:-M9YwbQyAr8mYy0ML1qJ:b-tlx7ko)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @asubiotto and @rytaft)

___
*[pkg/sql/distsql_spec_exec_factory.go, line 222 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/49857#-M9OEk_k0qyym0RVddfg:-M9Yw8WoEClBkKSqutwP:b6cmexm) ([raw file](https://github.com/cockroachdb/cockroach/blob/7aab3b0c352b1b23f073b54561aca7d988d4bccf/pkg/sql/distsql_spec_exec_factory.go#L222)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

This is what I mean:
```
gatewayNodeID := roachpb.NodeID(p.execCfg.NodeID.SQLInstanceID())
```
Similar to what I did in `explain_distsql.go`:
```
		// TODO(asubiotto): This cast from SQLInstanceID to NodeID is temporary:
		//  https://github.com/cockroachdb/cockroach/issues/49596
		flows := physPlan.GenerateFlowSpecs(roachpb.NodeID(params.extendedEvalCtx.NodeID.SQLInstanceID()))
```

It's a fair point that the `nodeDesc` already exists and this will just add an extra usage. I'll leave it up to you to decide what you want to do.
</blockquote></details>

Huh, I thought I tried this variant and the compiler was complaining. Thanks for explicitly typing it out, done.


<!-- Sent from Reviewable.io -->
